testng-core/src/main/java/org/testng/DataProviderHolder.java

```diff
@@ -38,7 +40,7 @@ public void addInterceptor(IDataProviderInterceptor interceptor) {
   }
 
   public void merge(DataProviderHolder other) {
-    this.listeners.addAll(other.getListeners());
+    addListeners(other.getListeners());
```

Comment:
Listeners in TestNG are always treated to be at a global level. In this case, we are finding the same listener in two `<test>` tags. So when we end up consolidating them, we end up having 2 listener instances (which is the duplicate). I am not sure how we can avoid this because avoiding this would mean that we would start having to associate scope for listeners (class level, `<test>` tag level) and that I feel is a larger change.

Reply:
I agree the change will be bigger. I think it is something we should really consider for the futur.

Reply:
I will add this as an entry to the discussion thread that I created, so that it's easier for me to start creating issues out of it later.

Reply:
Here's the link to the entry https://github.com/testng-team/testng/discussions/3037#discussioncomment-8414586